### ISSUE TITLE
Don't block pika's I/O loop

### ISSUE THREAD

**Issue by lukebakken** [saml/x#1]
This line blocks Pika's I/O loop, causing RabbitMQ to think that the connection is dead due to missed heartbeat:

https://github.com/saml/x/blob/master/pika_heartbeat/pika_heartbeat.py#L40

I am assuming that in your production code the `sleep()` is there to publish messages at a certain rate, is that true?

**Comment by saml** [comment]
@lukebakken `sleep()` is there to simulate idle producer. The producer only publishes message upon certain events in our system. And it may sit idle until the next event. And idle period can be over heartbeat (60 seconds).

**Comment by lukebakken** [comment]
Thanks for the explanation. It's the "sit idle" part in your production code that blocks Pika. Questions:

* I'm assuming that your production code is single-threaded. Is that true?
* What actually happens during the "sit idle" period in your production code?

**Comment by saml** [comment]
@lukebakken yes single threaded.
Nothing happens.

Are you trying to get at something? Is `sleep()` not a good emulation for sitting idle on a single threaded app?
I could easily replace sleep() with something else:
```py
ch.basic_publish("", "test", "")
# time.sleep(sleep_secs)
input("Press enter to continue")
ch.basic_publish("", "test", "")
```

What's a recommended way to "sit idle"?
Not sure why you're interested in production code.

I updated the example with more elaborate consumer that publishes to a different channel when a message is consumed:
https://github.com/saml/x/blob/347d5356e122034c069dd10472e0520200459090/pika_heartbeat/pika_heartbeat.py#L35-L52
You need to run rabbitmq and add a vhost and a queue. Or use the following command to set those up:
```
docker run --rm --publish '5672:5672' --publish '15672:15672' --volume $PWD/rabbitmq.json:/etc/rabbitmq/definitions.json rabbitmq:3.7-management-alpine
```
Afterwards, publish a message using management web UI to `test` queue. Wait a bit and publish to the queue again. The consumer will fail to publish to the different vhost cause broker closed publisher channel.

But this elaboration adds confusion. Actual test case is still:
1. Open connection and channel
2. Publish a message
3. Wait long enough before publishing another message
4. Observe failure during publish

Let me know a good way to "sit idle".

**Comment by woakesd** [comment]
Use the connection sleep method instead of timer.sleep.

https://pika.readthedocs.io/en/stable/modules/adapters/blocking.html#pika.adapters.blocking_connection.BlockingConnection.sleep

Ok reading the chain I see you aren't sleeping but waiting on the production of messages. I forked this code and have put a PR to modifictions similar to what I use, which is to have a timer firing and calling the connection sleep method. This allows the heartbeats to be sent to the RabbitMQ server and keeps the connection up. Feels clunky to be honest though as you need to use a threading.Lock to avoid calling connection sleep while publishing or vice versa.

**Comment by lukebakken** [comment]
The gist of all of this is that you **can't block Pika's I/O loop**. In a single threaded app, _any_ work that is done outside of Pika will block the loop. This is not unique to Pika but is how an I/O loop works in any single-threaded process.

You have these options:

* Use multiple threads, and use `connection.add_callback_threadsafe` to publish messages on Pika's thread.
* Use @woakesd 's solution and sleep correctly using `connection.sleep`.

**Comment by saml** [comment]
@lukebakken yeah, I can verify that even on consumer, if callback that handles each message does not return before heartbeat interval, server closes connection and consumer will die.

https://pika.readthedocs.io/en/stable/examples/heartbeat_and_blocked_timeouts.html  recommends to increase heartbeat interval.
> As a result, applications that perform lengthy processing in the same thread that also runs their Pika connection may experience unexpected dropped connections due to heartbeat timeout

An example of using multiple threads to keep up with heartbeat would be great (probably using `connection.add_callback_threadsafe` and `connection.sleep`: calling `connection.sleep` in different thread at an interval shorter than heartbeat). I'll try to implement one myself as well. Not sure why this multi-thread option to keep the BlockingConnection alive isn't included in the library. Seems to be an useful add-on that can be turned on optionally.

**Comment by lukebakken** [comment]
> Not sure why this multi-thread option to keep the BlockingConnection alive isn't included in the library.

It all comes down to availability of free time (or payment) to implement it. Pika's scope is very low-level at the moment.

I will put together a quick example for publishing on an interval in a PR.

**Comment by lukebakken** [comment]
Please see #4

**Comment by lukebakken** [comment]
Fixed by #4
